Introduce `cache-directory` config option to speed up analysis

ruudk opened the pull request:
Now there was already a discussion in #295 if this was needed or not and what is considered fast or not. For us, we spend a lot of time in optimizing our CI pipeline to let it finish within a few minutes. Shaving off 3 minutes is worth it for us. I can understand that you might not want to merge this in, and that's fine. I'll resort to my own fork then. 

## Description

We use `ComposerRequireChecker` for our monolith with 10k+ classes and run it on GitHub Actions.
The average run time is 5 minutes and 30 seconds. And that's only `ComposerRequireChecker`. Together with the job setup it's even slower.

By setting the `cache-directory` in our configuration file we are able to cache the parsed PHP files
and reuse them for the next run.

Now, the checker runs in 2 minute and 30 seconds. Making the total difference 5:30 => 2:00 = 3 minutes faster.

Ocramius commented on src/ComposerRequireChecker/ASTLocator/LocateASTFromFiles.php:
```diff
@@ -39,11 +45,21 @@ public function __invoke(Traversable $files): Traversable
+                    throw new FileParseFailed($file, $e);
+                }
+
+                $this->cache->set($key, $stmts);
```
This caches the parsed AST, not the found symbols (which is what this tool actually processes).

Because this cache will then contain third-party symbols (defined by `nikic/php-parser`), I think this approach is not applicable.

Ideally, we should parse the resulting found symbols, which is a data structure "owned" by this package.

ruudk replied:
It took some refactoring to be able to cache the symbols while only calling the parser on a cache miss. What do you think?

Ocramius commented on src/ComposerRequireChecker/Cli/CheckCommand.php:
```diff
@@ -260,12 +263,21 @@ private function getComposerData(string $jsonFile): array
+            return null;
+        }
+
+        return new Psr16Cache(new FilesystemAdapter('composer-require-checker', 7 * 86400, $options->getCacheDirectory()));
```
The `7 * 86400` is a bit arbitrary: would say "forever" instead?

ruudk replied:
The default is `0` (forever) but that's not ideal because it will mean that caches will never be cleaned up. When running this in a CI pipeline you'll end up with a very large cache. Therefore, we need to find a value that's not too short and not forever.

Ocramius replied:
`86400 * 7` is basically the same thing as "forever", except that it leads to weekly hiccups :-\

ruudk replied:
Changed to `forever`

Ocramius commented on src/ComposerRequireChecker/ASTLocator/LocateASTFromFiles.php:
```diff
@@ -39,11 +45,21 @@ public function __invoke(Traversable $files): Traversable
-            } catch (Error $e) {
-                // Convert the parse error into one which has the file information
-                throw new FileParseFailed($file, $e);
+            $key = sha1_file($file);
```
This hashing should occur one level up, when looking at directories: that would allow for easier cache slamming done implicitly by this tool.

The cache key should factor in the current `composer.lock` and `installed.json`, I'd say :thinking:

ruudk replied:
The collected symbols for a given file are always the same, unless the file was modified, right?
So if we have the symbols in cache based on the hashed file contents, it can never be different. Or do I miss something.

Ocramius replied:
Depends on whether the code in this package (or any dependency that has to do with this package) changes behavior in any way.

ruudk replied:
In the `SymbolCache` we can modify the `getKey` method and include other things.

ruudk committed:
Introduce `cache-directory` config option to speed up analysis

We use `ComposerRequireChecker` for our monolith with 10k+ classes and run it on GitHub Actions.
The average run time is 5 minutes and 30 seconds. And that's only `ComposerRequireChecker`. Together with the job setup it's even slower.

By setting the `cache-directory` in our configuration file we are able to cache the parsed PHP files
and reuse them for the next run.

Now, the checker runs in 2 minute and 30 seconds. Making the total difference 5:30 => 2:00 = 3 minutes faster.

Now there was already a discussion in #295 if this was needed or not and what is considered fast or not. For us, we spend a lot of time in optimizing our CI pipeline to let it finish within a few minutes. Shaving off 3 minutes is worth it for us. I can understand that you might not want to merge this in, and that's fine. I'll resort to my own fork then.

ruudk commented:
Refactored the PR. It now caches the symbols instead of the AST. Looking forward to your review.

Ocramius commented:
Was this closed on purpose?

ruudk commented:
Yes, I closed it because of this finding: https://github.com/maglnet/ComposerRequireChecker/issues/295#issuecomment-997769794

If you still think the cache is worth it I can reopen it.

Ocramius commented:
Ah, thanks for clarifying! Indeed, xdebug + `nikic/php-parser` is quite a hit :|
